### ISSUE TITLE
let `return 0` on toplevel return exit status like bash instead of fatal error

### ISSUE THREAD

**Issue by bar-g** [oilshell/oil#1715]
Replacement issue for #1696, to not loose the detected deviation from bash behavior.

 Found by trying a bash method to detect if the script is being run directly or sourced:
`(return 0 2>/dev/null) && sourced=1 || sourced=0`

(That method as posted at https://stackoverflow.com/questions/2683279/how-to-detect-if-a-script-is-being-sourced
is basicly expected to work in all shells where `return 0` fails with exit code >0 if called on top level (e.g. bash), while avoiding the specific bash variables and possibly pitfalls of `[ "$0" = "$BASH_SOURCE" ]`. The method is also related to "exit instead of returning, if executed directly": `return 2>/dev/null || exit`)

However, in osh the error(s) seem fatal. Example:
```
osh-0.17.0$ (return 0)
  (return 0)
   ^~~~~~
[ interactive ]:1: fatal: Invalid control flow 'return' in pipeline / subshell / background
```

EDIT: split-off osh error with "forkwait { true }", and ysh error message related to parsing "if forkwait"

----
Does ysh maybe support a better/direct/obvious/readable method to determine if code being sourced?
Would it be worthwhile?

----

Solution:

- [ ]  let `return` exit with status >0 if called in top level (bash behavior), instead of throwing a fatal error. (Or let it auto-fallback to do an exit like other shells???)

**Comment by bar-g** [comment]
Ah, a reason for the fatal error, then that's something I haven't understood yet.

**Comment by andychu** [comment]
Looking at that link again, it apparently only works in bash.

I think I disallowed that semantic because it just seemed like undocumented bash  quirks.  Did bash **document it** anywhere?

bash changes undocumented behavior from version to version -- I've encountered many instances of it.

---

It looks like you already have to detect the shell anyway, so I guess a solution is to detect `$OILS_VERSION` and use `if is-main`

**Comment by bar-g** [comment]
Yes, concerning the "if sourced/main" implementations, I think the only more or less portable method was rather in this direction:`if [ ${_sourced_file:-${BASH_SOURCE:-hardcodedfile}} = ${_main_file:-${$0}} ], i.e. not using `return 0` at all.

I'm not sure if a portable alternative for "exit instead of returning, if executed directly" exists `return 2>/dev/null || exit` exists. 

It seems documented though:
man bash:
>  If return  is  used  outside  a function, but during execution of a
              script by the .  (source) command, it causes the shell  to  stop
              executing  that script and return either n or the exit status of
              the last command executed within the script as the  exit  status
              of  the script.  If n is supplied, the return value is its least
              significant 8 bits.  The **return status is non-zero if return**  is
              supplied  a  non-numeric argument, or **is used outside a function
              and not during execution of a script by . or source**.   Any  com‐
              mand  associated  with the RETURN trap is executed before execu‐
              tion resumes after the function or script.
